docs/CppInteroperability/SwiftToCppInteroperabilityRoadmap.md

```diff
@@ -0,0 +1,125 @@
+
+### Performance
+
+Swift to C++ bridging should be as efficient as possible. The Swift compiler should avoid unnecessary overhead for calling into Swift functions from C++, or using Swift types from C++. Furthermore, the bridging code should not convert Swift types into their C++ counterparts automatically as that can add a lot of overhead. For instance, a Swift function returning a String should still return a Swift String in C++, and not a std::string.
+
+Certain Swift features do not have a corresponding C++ mapping which can impose some additional overhead to make them work in C++. Resilient value types are a good example of this, as their layout and size is not known at compile time. This means that the generated C++ type that represents such a Swift type might need to impose additional performance overhead by allocating the storage for such value dynamically on the heap (i.e. boxing). In cases like this, the Swift compiler should strive to have minimal possible overhead, by finding solutions where this overhead can be optimized out for common cases.
+
+### Achieving Safety With Performance In Mind
+
+Certain aspects of Swift’s memory model impose certain restrictions that create tension between the goal of achieving safety and the goal of avoiding unnecessary overhead for calling into Swift from C++. Checking for exclusivity violations is a good example of this, as the C++ compiler does not have a notion of exclusivity it can verify, meaning that it is difficult to prove that a value is accessed exclusively in the C++ code that calls into Swift. This means that the C++ code that calls into Swift APIs will most likely require more run-time checks to validate exclusivity than similar Swift code that calls the same Swift APIs.
+
+The adherence to Swift’s type and memory safety rules should be prioritized ahead of performance when C++ calls into Swift, even if this means more run-time checks are required. This is needed to avoid unexpected crashes or other unexpected behavior in users Swift code. For users seeking maximum performance, Swift provides additional compiler flags that avoid certain run-time checks. Those flags should be taken into account when the C++ header is generated. An example of such flag is -enforce-exclusivity. When  -enforce-exclusivity=none is passed to Swift, the Swift compiler does not emit any run-time checks that check for exclusivity violations. A flag like this should also affect the generated C++ header, and the Swift compiler should not emit any run-time checks for exclusivity in the generated C++ header when this flag is used.
```

Comment:
Yes, the flags are honored because the thunks are emitted inline in the header by the Swift compiler. So Swift can adjust the code in the thunk based on the flags that are used when the header is generated.